code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
-        uint8_t receive_byte() {
-            return hardware_usart->US_RHR;
+        constexpr static uint8_t receive_byte() {
+            return detail::usart::port<Bus>->US_RHR;
         }
 
-        void set_peripheral(Pio *pio, uint32_t mask, peripheral p) {
-            uint32_t t = pio->PIO_ABSR;
+        template <uint32_t mask>
+        constexpr void set_peripheral() {
+            uint32_t t = detail::pio::port<typename Bus::pio>->PIO_ABSR;
 
-            if (p == peripheral::peripheral_a) {
-                pio->PIO_ABSR &= (~mask & t);
+            if constexpr (std::is_same_v<typename Bus::periph,
+                                         detail::pio::periph_a>) {
+                detail::pio::port<typename Bus::pio>->PIO_ABSR = (~mask & t);
             } else {
-                pio->PIO_ABSR = (mask | t);
+                detail::pio::port<typename Bus::pio>->PIO_ABSR = (mask | t);
             }
 
             // remove pin from pio controller
-            pio->PIO_PDR = mask;            
+            detail::pio::port<typename Bus::pio>->PIO_PDR = mask;
+        };
+
+        constexpr void enable() {
+            detail::usart::port<Bus>->US_CR = UART_CR_RXEN | UART_CR_TXEN;
+        }
+
+        constexpr void disable() {
+            detail::usart::port<Bus>->US_CR =
+                UART_CR_RSTRX | UART_CR_RSTTX | UART_CR_RXDIS | UART_CR_TXDIS;
         }
 
     public:
-        hardware_usart_c(unsigned int baudrate, usart_ports usart_port) {
-            if (usart_port == usart_ports::UART_SIZE){
-                HWLIB_PANIC_WITH_LOCATION;
-            }
+        hardware_usart_c(unsigned int baudrate) {
+            // set the peripheral for the usart pins
+            set_peripheral<Bus::rx>();
+            set_peripheral<Bus::tx>();
+
+            // enable the peripheral clock for the usart
+            PMC->PMC_PCER0 = (0x01 << detail::periph_id::id<Bus>);
+
+            // disable all interrupts for the usart
+            detail::usart::port<Bus>->US_IDR = 0xFFFFFFFF;
 
-            const auto &curr = usart[uint8_t(usart_port)];
+            // clear the status register for old interrupts
+            (void)detail::usart::port<Bus>->US_CSR;
 
-            hardware_usart = curr.usart;
-            set_peripheral(curr.pio, curr.rx_mask, curr.periph);
-            set_peripheral(curr.pio, curr.tx_mask, curr.periph);
-            PMC->PMC_PCER0 = (0x01 << curr.id);
+            // use the periph_id since that is the same id as the IRQn_type
+            constexpr auto iqrn =
+                static_cast<IRQn_Type>(detail::periph_id::id<Bus>);
+
+            // enable the nvic interrupt for the current usart
+            NVIC_SetPriority(iqrn, 9);
+            NVIC_EnableIRQ(iqrn);
 
             disable();
 
-            hardware_usart->US_BRGR = (5241600u / baudrate);
+            // write the baud rate in the BRGR register
+            detail::usart::port<Bus>->US_BRGR = (5241600u / baudrate);
```

Comment:
What happens here if two modules chose the same bus? Does the second instantiation decide the baudrate?

Reply:
Currently yes

Reply:
I think it would be wise to add this to the wiki/readme if it isn't already, I can imagine it creating unforseen issues for users of the library working on the same due

Reply:
I have added this as a @warning in the doxygen comments.

Reply:
I have added this to the readme